collector/cache.go

```diff
@@ -0,0 +1,453 @@
+package collector
+
+import (
+	"github.com/StackExchange/wmi"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+func init() {
+	Factories["cache"] = NewCacheCollector
+}
+
+// A CacheCollector is a Prometheus collector for WMI Win32_PerfFormattedData_PerfOS_Cache  metrics
+type CacheCollector struct {
+	AsyncCopyReadsPersec         *prometheus.Desc
+	AsyncDataMapsPersec          *prometheus.Desc
+	AsyncFastReadsPersec         *prometheus.Desc
+	AsyncMDLReadsPersec          *prometheus.Desc
+	AsyncPinReadsPersec          *prometheus.Desc
+	CopyReadHitsPercent          *prometheus.Desc
+	CopyReadsPersec              *prometheus.Desc
+	DataFlushesPersec            *prometheus.Desc
+	DataFlushPagesPersec         *prometheus.Desc
+	DataMapHitsPercent           *prometheus.Desc
+	DataMapPinsPersec            *prometheus.Desc
+	DataMapsPersec               *prometheus.Desc
+	DirtyPages                   *prometheus.Desc
+	DirtyPageThreshold           *prometheus.Desc
+	FastReadNotPossiblesPersec   *prometheus.Desc
+	FastReadResourceMissesPersec *prometheus.Desc
+	FastReadsPersec              *prometheus.Desc
+	LazyWriteFlushesPersec       *prometheus.Desc
+	LazyWritePagesPersec         *prometheus.Desc
+	MDLReadHitsPercent           *prometheus.Desc
+	MDLReadsPersec               *prometheus.Desc
+	PinReadHitsPercent           *prometheus.Desc
+	PinReadsPersec               *prometheus.Desc
+	ReadAheadsPersec             *prometheus.Desc
+	SyncCopyReadsPersec          *prometheus.Desc
+	SyncDataMapsPersec           *prometheus.Desc
+	SyncFastReadsPersec          *prometheus.Desc
+	SyncMDLReadsPersec           *prometheus.Desc
+	SyncPinReadsPersec           *prometheus.Desc
+}
+
+// NewCacheCollector ...
+func NewCacheCollector() (Collector, error) {
+	const subsystem = "cache"
+	return &CacheCollector{
+		AsyncCopyReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "async_copy_reads_persec"),
```

Comment:
There's a lot of these `_persec` metrics here, which from a look at one of my systems seem to be named differently from their actual values, because their names are from a calculated value that is displayed in Performance Monitor. The values we read are actually counters, with the total number of operations.